core/shared/src/main/scala/zio/internal/FiberContext.scala

```diff
@@ -412,15 +415,18 @@ private[zio] final class FiberContext[E, A](
                 case ZIO.Tags.EffectAsync =>
                   val zio = curZio.asInstanceOf[ZIO.EffectAsync[Any, E, Any]]
 
+                  val epoch = asyncEpoch
+                  asyncEpoch = epoch + 1
```

Comment:
Just `asyncEpoch += 1` should be ok

Reply:
need to capture current value to local `val`
and avoid double volatile reads